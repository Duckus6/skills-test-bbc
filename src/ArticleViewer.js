import React, {useState} from 'react';
import './App.css';
import ArticleAPI from './ArticleAPI';
import Article from './Article';

function ArticleViewer() {
    const API = new ArticleAPI();
    const [articles,setArticles] = useState([]);
    React.useEffect(()=>{
        API.getArticles().then(
            articles=>{
                setArticles(articles);
                console.debug(`${articles.length} articles saved to state`)
            }
        ).catch(
            err=>console.log(err)
        );
    },[]);

    const [articleIndex,setArticleIndex] = useState(0);


    return (
        <div className="ArticleViewer">
            {articles &&<p>We have {articles.length} Articles</p>}
            {((articles) && (articles.length>0 )) && <Article articleID={articles[articleIndex]}/>}
        </div>
    );
}

export default ArticleViewer;
