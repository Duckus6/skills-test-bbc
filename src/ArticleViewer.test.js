import React from 'react';
import AxiosMockAdapter from 'axios-mock-adapter';
import {mockArticle,mockArticles,mockError} from "./fixtures/articles";
import {shallow} from "enzyme";
import ArticleViewer from "./ArticleViewer";
import axios from "axios/index";
import {act} from "@testing-library/react";




const mock = new AxiosMockAdapter(axios);



describe("the server returns 200 code",()=>{
    beforeEach(()=>{
        mock.onGet('data/').reply(200,
            mockArticles
        );
        mock.onGet(`data/${mockArticles[0]}`).reply(200,
            mockArticle
        );
        jest.spyOn(React,"useEffect").mockImplementationOnce(f=>f());
    });
    afterEach(()=>{
        mock.reset();
    });
    test('renders', async (done) => {
        jest.useFakeTimers();
        const wrapper = shallow(<ArticleViewer/>);
        await act(async ()=>{
            jest.runAllImmediates();
        });
        wrapper.update();
        console.log(wrapper.debug());
        expect(wrapper.debug()).toMatchSnapshot();
        done();
    })
});

describe("the server returns 500 error",()=>{
    beforeEach(()=>{
        mock.onGet('data/').reply(500,
            mockError
        );
        jest.spyOn(React,"useEffect").mockImplementationOnce(f=>f());
    });
    afterEach(()=>{
        mock.reset();
    });
    test('renders', () => {
        const wrapper = shallow(<ArticleViewer/>);
        expect(wrapper.debug()).toMatchSnapshot();
    })
});
