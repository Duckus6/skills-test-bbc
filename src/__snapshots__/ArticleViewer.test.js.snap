// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the server returns 200 code renders 1`] = `
"<div className=\\"ArticleViewer\\">
  <p>
    We have 
    5
     Articles
  </p>
  <Article articleID=\\"article1.json\\" />
</div>"
`;

exports[`the server returns 500 error renders 1`] = `
"<div className=\\"ArticleViewer\\">
  <p>
    We have 
    0
     Articles
  </p>
</div>"
`;
